### ISSUE TITLE
Gradient of SVD is NaN if two singular values are equal

### ISSUE THREAD

**Issue by JanLuca** [google/jax#8732]
Calculating a gradient of the SVD of a matrix where at least two of the singular values are equal lead to a gradient with only NaNs. Let me illustrate the problem by a sample code and result:

```python
import jax
import jax.numpy as jnp

def f(arr):
    U, S, Vh = jnp.linalg.svd(arr, full_matrices=False)
    return jnp.linalg.norm((U * S) @ Vh)

A = jnp.diag(jnp.asarray([1, 1, 2, 3], dtype=float))

f_g = jax.value_and_grad(f)

print(f_g(A))
```
Result:
```python
(DeviceArray(3.8729835, dtype=float32),
 DeviceArray([[nan, nan, nan, nan],
              [nan, nan, nan, nan],
              [nan, nan, nan, nan],
              [nan, nan, nan, nan]], dtype=float32))
```
This sample is nothing but a very complicated expression of the norm of a matrix whose gradient is defined (easy to see if replacing the gradient call by `jax.value_and_grad(jnp.linalg.norm)` [2]).

After some discussion with @frederikwilde (the author of the commit f443e19141c3d04a6e2947a0dcde8d23039af786) we found that the problem is in the calculation of the derivative of the singular matrices _U_ and _V<sup>&dagger;</sup>_. The derivation of the derivative calculates in one step the matrix of the inverses _1/(S<sub>i</sub> - S<sub>j</sub>)_ (with _S<sub>i</sub>_ the singular values). Obviously, this leads to the described problem.

In the source code https://github.com/google/jax/blob/main/jax/_src/lax/linalg.py#L1291 there is already the commented out code part [1] which would implement a pseudo-inverse in this case. As far as I understand the reason why this code is not used at the moment is the general problem that there is a gauge-freedom of the singular subspace (which has n>1 if there are two equal singular values) is not solved and therefore the derivative of only one of the two unitaries alone would make no sense in this case.

As far as I understand the problem, it should be ok to do the pseudo-inverse of the differences, if users are not interested in the derivative of a single unitary but of the complete product _U * S @ Vh_ or variants of this one. To illustrate that enabling the code path [1] leads to correct results in some cases, one can replace the code line and rerun the example above. The resulting gradient is the same one as if one used the VJP-rule of the `jnp.linalg.norm` function. [3]

What is your opinion on this problem? Can the gauge-freedom somehow be fixed so it is no problem? Is the use case of using only the derivative of one unitary very unlikely so one could stick to the pseudo-inverse?

[1] `s_diffs_zeros = jnp.ones((), dtype=s.dtype) * (s_diffs == 0.)` (fixing the bug that it has to be `s.dtype` instead of `A.dtype`)

[2]
```python
import jax
import jax.numpy as jnp

A = jnp.diag(jnp.asarray([1, 1, 2, 3], dtype=float))

norm_g = jax.value_and_grad(jnp.linalg.norm)

print(f_g(A))
```
Result:
```python
(DeviceArray(3.8729832, dtype=float32),
 DeviceArray([[0.2581989, 0.       , 0.       , 0.       ],
              [0.       , 0.2581989, 0.       , 0.       ],
              [0.       , 0.       , 0.5163978, 0.       ],
              [0.       , 0.       , 0.       , 0.7745967]], dtype=float32))
```

[3] Result for example code with https://github.com/google/jax/blob/main/jax/_src/lax/linalg.py#L1291 replaced by [1]:
```python
(DeviceArray(3.8729835, dtype=float32),
 DeviceArray([[0.25819886, 0.        , 0.        , 0.        ],
              [0.        , 0.25819886, 0.        , 0.        ],
              [0.        , 0.        , 0.5163977 , 0.        ],
              [0.        , 0.        , 0.        , 0.7745966 ]], dtype=float32))
```

**Comment by shoyer** [comment]
There is a similar issue for calculating gradients through eigenvalue problems with degeneracies: https://github.com/google/jax/issues/669

Fundamentally, I think the problem is that `svd` and `eigh` are not well-defined functions (due to gauge freedom) in the case of degeneracies. But on the other hand, in practice these functions are always being used to calculated well-defined quantities when gradients are being calculated.

So from my perspective, there are two options:
1. Define derivative rules that are always correct in the case of degeneracies, when being used to calculate some downstream well-defined quantity. I think this may be possible but I'm not entirely sure. My preference would be for someone to write up  _formal proof_, at least in the form of a pre-print.
2. Encourage users to write custom derivative rules (maybe by providing some helper functions?) for the downstream functions that use SVD/Eigendecomposition in the case of degeneracies. I suspect that a few patterns (e.g., matrix valued functions) could capture most of these use-cases.

**Comment by jakevdp** [comment]
Closing as a duplicate of #2311

**Comment by EmilianoG-byte** [comment]
Hi @JanLuca, it's been a while since this issue was opened, but I am having the same issue and the linke that @jakevdp provided (#2311) did not solve my problem, unfortunately.

Could you maybe re-link the lines that you are refferring to in your explanation? Currently they point to some unrelated code.

Additionally, having a more explicit explanation on how to apply this pseudo-inverse would be very helpful! I am currently implementing a custom backward rule but it's not accurate enough.

Thank you!

**Comment by tatopenn-cell** [comment]
Note up front: our fix targets `jnp.linalg.eigh`'s gradient rule specifically, not `svd` directly -- this doesn't close this issue, which is about SVD. Posting in case the same masking approach is a useful starting point for whoever tackles the SVD case, since the failure mode is the same shape (`1/(sigma_i - sigma_j)` for SVD vs `1/(lambda_i - lambda_j)` for eigh, both blowing up at degeneracy).

@shoyer's option 2 on the linked #2311 (custom derivative rules for downstream well-defined quantities) is what we ended up shipping.

We needed `jax.grad` through Uhlmann fidelity `F(rho_A, rho_B) = Tr(sqrt(sqrt(rho_A) rho_B sqrt(rho_A)))^2` between density matrices in a quantum-circuit simulator, where `rho_A` is frequently (near-)degenerate in practice -- e.g. the fully mixed state, or any noisy near-pure state with several near-zero, near-degenerate eigenvalues. `jnp.linalg.eigh`'s built-in gradient rule divides by `lambda_i - lambda_j` and returns NaN exactly there.

Fix is the practical masking variant mentioned in Kasim, "Derivatives of partial eigendecomposition of a real symmetric matrix for degenerate cases" (arXiv:2011.04366): a `jax.custom_jvp` on `eigh` that masks the `1/(lambda_i - lambda_j)` term to 0 for near-degenerate pairs instead of letting it blow up. Not Kasim's fuller per-degenerate-block treatment (which recovers a generally nonzero contribution from within the degenerate eigenspace itself for some perturbation directions) -- sufficient here because the only downstream use of the eigenvectors is rebuilding a matrix square root via its trace, so a masked (if not maximally sharp) subgradient is still mathematically valid.

**Edit**: I initially wrote below that this masking approach was "also used by PyTorch's `torch.linalg.eigh` gradient notes and the `xitorch` library's `symeig`" -- that was wrong, I hadn't actually checked, and I should have. Having now read the real source: PyTorch's current `linalg_eig_jvp`/`linalg_eig_backward` (`torch/csrc/autograd/FunctionsManual.cpp`) divides directly by `(L_j - L_i)` with no degenerate-pair masking at all -- same singularity as JAX's default, unresolved. xitorch's own derivation notes (`doc/notes/deriv_symeig.rst`, by the same Kasim) state up front "This derivation assumes the eigenvalues are all unique. Cases with degenerate eigenvalues are treated differently" without giving that treatment on that page, and `symeig`'s own docstring warns directly: "If `s` contains very small numbers or degenerate values, the calculation and its gradient might be inaccurate." So this appears to still be an open problem across the ecosystem, not something already solved elsewhere that we ported -- correcting the record rather than leave an unverified claim standing.

Source + full docstring: https://github.com/tatopenn-cell/Dense-Evolution/blob/1cb93a61b71a185222618601efb220d327e6998a/dense_evolution/mitigation.py#L303-L347

Verified two ways: (1) directional-derivative/JVP comparison against symmetric-tangent finite differences (avoids eigenvector sign/ordering ambiguity) -- exact match, diff=0.000000, on non-degenerate, 2-fold, and 3-fold degenerate test matrices; (2) the fully-mixed-state gradient (worst case: every eigenvalue tied) goes from all-NaN to finite, confirmed directly rather than assumed from the general theory.
